cpp/tensorrt_llm/nanobind/executor/executor.cpp

```diff
@@ -0,0 +1,239 @@
+        kind = 'f'; // floating point
+        break;
+    case static_cast<uint8_t>(nb::dlpack::dtype_code::Bfloat):
+        kind = 'f'; // brain floating point (treat as float kind)
+        break;
+    case static_cast<uint8_t>(nb::dlpack::dtype_code::Complex):
+        kind = 'c'; // complex
+        break;
+    default:
+        kind = 'V'; // void/other
+        break;
+    }
+    tle::DataType dtype;
+    if (npDtype == nb::dtype<half>())
+    {
+        dtype = tle::DataType::kFP16;
+    }
+    else if (npDtype == nb::dtype<float>())
+    {
+        dtype = tle::DataType::kFP32;
+    }
+    else if (npDtype == nb::dtype<int8_t>())
+    {
+        dtype = tle::DataType::kINT8;
+    }
+    else if (npDtype == nb::dtype<int32_t>())
+    {
+        dtype = tle::DataType::kINT32;
+    }
+    else if (npDtype == nb::dtype<int64_t>())
+    {
+        dtype = tle::DataType::kINT64;
+    }
+    else if (kind == 'V' && array.itemsize() == 1)
+    {
+        dtype = tle::DataType::kFP8;
+    }
+    else if (kind == 'V' && array.itemsize() == 2)
+    {
+        dtype = tle::DataType::kBF16;
+    }
+    else
+    {
+        TLLM_THROW("Unsupported numpy dtype.");
+    }
+
+    // todo: improve the following code
+    std::vector<int64_t> dims;
+    for (size_t i = 0; i < array.ndim(); ++i)
+    {
+        dims.push_back(static_cast<int64_t>(array.shape(i)));
+    }
+    tle::Shape shape(dims.data(), dims.size());
+
+    return tle::Tensor::of(dtype, const_cast<void*>(array.data()), shape);
+}
+
+} // namespace
+
+namespace tensorrt_llm::nanobind::executor
+{
+
+Executor::Executor(
+    std::filesystem::path const& modelPath, tle::ModelType modelType, tle::ExecutorConfig const& executorConfig)
+{
+    mExecutor = std::make_unique<tle::Executor>(modelPath, modelType, executorConfig);
+}
+
+Executor::Executor(std::filesystem::path const& encoderModelPath, std::filesystem::path const& decoderModelPath,
+    tle::ModelType modelType, tle::ExecutorConfig const& executorConfig)
+{
+    mExecutor = std::make_unique<tle::Executor>(encoderModelPath, decoderModelPath, modelType, executorConfig);
+}
+
+Executor::Executor(nb::bytes const& engineBuffer, std::string const& jsonConfigStr, tle::ModelType modelType,
+    tle::ExecutorConfig const& executorConfig, std::optional<nb::dict> managedWeights)
+{
+    uint8_t const* data = static_cast<uint8_t const*>(engineBuffer.data());
+    size_t size = engineBuffer.size();
+    std::optional<std::map<std::string, tle::Tensor>> managedWeightsMap = std::nullopt;
+    if (managedWeights.has_value() && !managedWeights.value().empty())
+    {
+        managedWeightsMap = std::map<std::string, tle::Tensor>();
+        for (auto const& item : managedWeights.value())
+        {
+            std::string name = nb::cast<std::string>(item.first);
+            nb::ndarray<nb::numpy> array = nb::cast<nb::ndarray<nb::numpy>>(item.second);
+            managedWeightsMap->emplace(name, numpyToTensor(array));
+        }
```

Comment:
```suggestion    
        for (auto const& [rawName, rawArray] : managedWeights.value())
        {
            std::string name = nb::cast<std::string>(rawName);
            nb::ndarray<nb::numpy> array = nb::cast<nb::ndarray<nb::numpy>>(rawArray);
            managedWeightsMap->emplace(name, numpyToTensor(array));
        }
```
Nitpick: Structured bindings can be clearer than first and second